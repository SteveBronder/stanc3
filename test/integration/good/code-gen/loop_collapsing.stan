data {
    int N;
    vector[N] d_x;
    int indexer[N];
}

transformed data {
    vector[N] td_x;
    for (i in 1:N) {
        td_x[i] = exp(d_x[i]);
    }
    td_x[:] = d_x[:];
    for (i in 1:N) {
        td_x[indexer[i]] = exp(d_x[indexer[i]]);
    }
}