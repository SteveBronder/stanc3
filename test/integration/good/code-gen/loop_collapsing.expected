  $ ../../../../../install/default/bin/stanc --debug-transformed-mir loop_collapsing.stan
((functions_block ())
 (input_vars
  ((N SInt)
   (d_x
    (SVector
     ((pattern (Var N))
      (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
   (indexer
    (SArray SInt
     ((pattern (Var N))
      (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
 (prepare_data
  (((pattern
     (Decl (decl_adtype DataOnly) (decl_id pos__) (decl_type (Sized SInt))))
    (meta <opaque>))
   ((pattern
     (Assignment (pos__ UInt ())
      ((pattern (Lit Int 1))
       (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
    (meta <opaque>))
   ((pattern
     (Decl (decl_adtype DataOnly) (decl_id N) (decl_type (Sized SInt))))
    (meta <opaque>))
   ((pattern
     (Assignment (N UInt ())
      ((pattern
        (Indexed
         ((pattern
           (FunApp (CompilerInternal FnReadData)
            (((pattern (Lit Str N))
              (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
          (meta ((type_ (UArray UInt)) (loc <opaque>) (adlevel DataOnly))))
         ((Single
           ((pattern (Lit Int 1))
            (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
       (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
    (meta <opaque>))
   ((pattern
     (NRFunApp (CompilerInternal FnValidateSize)
      (((pattern (Lit Str d_x))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Lit Str N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Var N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
    (meta <opaque>))
   ((pattern
     (Decl (decl_adtype DataOnly) (decl_id d_x)
      (decl_type
       (Sized
        (SVector
         ((pattern (Var N))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))))
    (meta <opaque>))
   ((pattern
     (Block
      (((pattern
         (Decl (decl_adtype AutoDiffable) (decl_id d_x_flat__)
          (decl_type (Unsized (UArray UReal)))))
        (meta <opaque>))
       ((pattern
         (Assignment (d_x_flat__ (UArray UReal) ())
          ((pattern
            (FunApp (CompilerInternal FnReadData)
             (((pattern (Lit Str d_x))
               (meta ((type_ UVector) (loc <opaque>) (adlevel DataOnly)))))))
           (meta ((type_ (UArray UReal)) (loc <opaque>) (adlevel DataOnly))))))
        (meta <opaque>))
       ((pattern
         (Assignment (pos__ UInt ())
          ((pattern (Lit Int 1))
           (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
        (meta <opaque>))
       ((pattern
         (For (loopvar sym1__)
          (lower
           ((pattern (Lit Int 1))
            (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
          (upper
           ((pattern (Var N))
            (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
          (body
           ((pattern
             (Block
              (((pattern
                 (Assignment
                  (d_x UVector
                   ((Single
                     ((pattern (Var sym1__))
                      (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
                  ((pattern
                    (Indexed
                     ((pattern (Var d_x_flat__))
                      (meta
                       ((type_ (UArray UReal)) (loc <opaque>)
                        (adlevel DataOnly))))
                     ((Single
                       ((pattern (Var pos__))
                        (meta
                         ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
                   (meta ((type_ UReal) (loc <opaque>) (adlevel DataOnly))))))
                (meta <opaque>))
               ((pattern
                 (Assignment (pos__ UInt ())
                  ((pattern
                    (FunApp (StanLib Plus__ FnPlain)
                     (((pattern (Var pos__))
                       (meta
                        ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
                      ((pattern (Lit Int 1))
                       (meta
                        ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
                   (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
                (meta <opaque>)))))
            (meta <opaque>)))))
        (meta <opaque>)))))
    (meta <opaque>))
   ((pattern
     (NRFunApp (CompilerInternal FnValidateSize)
      (((pattern (Lit Str indexer))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Lit Str N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Var N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
    (meta <opaque>))
   ((pattern
     (Decl (decl_adtype DataOnly) (decl_id indexer)
      (decl_type
       (Sized
        (SArray SInt
         ((pattern (Var N))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))))
    (meta <opaque>))
   ((pattern
     (Assignment (indexer (UArray UInt) ())
      ((pattern
        (FunApp (CompilerInternal FnReadData)
         (((pattern (Lit Str indexer))
           (meta ((type_ (UArray UInt)) (loc <opaque>) (adlevel DataOnly)))))))
       (meta ((type_ (UArray UInt)) (loc <opaque>) (adlevel DataOnly))))))
    (meta <opaque>))
   ((pattern
     (NRFunApp (CompilerInternal FnValidateSize)
      (((pattern (Lit Str td_x))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Lit Str N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
       ((pattern (Var N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
    (meta <opaque>))
   ((pattern
     (Decl (decl_adtype DataOnly) (decl_id td_x)
      (decl_type
       (Sized
        (SVector
         ((pattern (Var N))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))))
    (meta <opaque>))
   ((pattern
     (For (loopvar i)
      (lower
       ((pattern (Lit Int 1))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
      (upper
       ((pattern (Var N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
      (body
       ((pattern
         (Block
          (((pattern
             (Assignment
              (td_x UVector
               ((Single
                 ((pattern (Var i))
                  (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
              ((pattern
                (FunApp (StanLib exp FnPlain)
                 (((pattern
                    (Indexed
                     ((pattern (Var d_x))
                      (meta
                       ((type_ UVector) (loc <opaque>) (adlevel DataOnly))))
                     ((Single
                       ((pattern (Var i))
                        (meta
                         ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
                   (meta ((type_ UReal) (loc <opaque>) (adlevel DataOnly)))))))
               (meta ((type_ UReal) (loc <opaque>) (adlevel DataOnly))))))
            (meta <opaque>)))))
        (meta <opaque>)))))
    (meta <opaque>))
   ((pattern
     (Assignment (td_x UVector (All))
      ((pattern
        (Indexed
         ((pattern (Var d_x))
          (meta ((type_ UVector) (loc <opaque>) (adlevel DataOnly))))
         (All)))
       (meta ((type_ UVector) (loc <opaque>) (adlevel DataOnly))))))
    (meta <opaque>))
   ((pattern
     (For (loopvar i)
      (lower
       ((pattern (Lit Int 1))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
      (upper
       ((pattern (Var N))
        (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))
      (body
       ((pattern
         (Block
          (((pattern
             (Assignment
              (td_x UVector
               ((Single
                 ((pattern
                   (Indexed
                    ((pattern (Var indexer))
                     (meta
                      ((type_ (UArray UInt)) (loc <opaque>)
                       (adlevel DataOnly))))
                    ((Single
                      ((pattern (Var i))
                       (meta
                        ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
                  (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
              ((pattern
                (FunApp (StanLib exp FnPlain)
                 (((pattern
                    (Indexed
                     ((pattern (Var d_x))
                      (meta
                       ((type_ UVector) (loc <opaque>) (adlevel DataOnly))))
                     ((Single
                       ((pattern
                         (Indexed
                          ((pattern (Var indexer))
                           (meta
                            ((type_ (UArray UInt)) (loc <opaque>)
                             (adlevel DataOnly))))
                          ((Single
                            ((pattern (Var i))
                             (meta
                              ((type_ UInt) (loc <opaque>)
                               (adlevel DataOnly))))))))
                        (meta
                         ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))))
                   (meta ((type_ UReal) (loc <opaque>) (adlevel DataOnly)))))))
               (meta ((type_ UReal) (loc <opaque>) (adlevel DataOnly))))))
            (meta <opaque>)))))
        (meta <opaque>)))))
    (meta <opaque>))))
 (log_prob ())
 (generate_quantities
  (((pattern
     (IfElse
      ((pattern
        (FunApp (StanLib PNot__ FnPlain)
         (((pattern
            (EOr
             ((pattern (Var emit_transformed_parameters__))
              (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
             ((pattern (Var emit_generated_quantities__))
              (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
           (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
       (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
      ((pattern (Block (((pattern (Return ())) (meta <opaque>)))))
       (meta <opaque>))
      ()))
    (meta <opaque>))
   ((pattern
     (IfElse
      ((pattern
        (FunApp (StanLib PNot__ FnPlain)
         (((pattern (Var emit_generated_quantities__))
           (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
       (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
      ((pattern (Block (((pattern (Return ())) (meta <opaque>)))))
       (meta <opaque>))
      ()))
    (meta <opaque>))))
 (transform_inits
  (((pattern
     (Decl (decl_adtype DataOnly) (decl_id pos__) (decl_type (Sized SInt))))
    (meta <opaque>))
   ((pattern
     (Assignment (pos__ UInt ())
      ((pattern (Lit Int 1))
       (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))))
    (meta <opaque>))))
 (output_vars ()) (prog_name loop_collapsing_model)
 (prog_path loop_collapsing.stan))
